CASSANDRA-17621 May 2022 blog "Apache Cassandra 4.1 Features: Guardrails Framework" (#128)
apache/cassandra-website: site-content/source/modules/ROOT/pages/blog/Apache-Cassandra-4.1-Features-Guardrails-Framework.adoc

In Apache Cassandra 4.1.0, we are introducing a new framework called Guardrails. The framework helps operators avoid certain configuration and usage pitfalls that can degrade the performance and availability of an Apache Cassandra cluster when taken to scale. 

For example, on the schema side, users can create too many tables or secondary indexes, leading to excessive use of resources. On the query side, users can run queries touching too many partitions that might involve all nodes in the cluster. Even worse, they can simply run a query using costly replica-side filtering, potentially reading all the table contents into memory on all nodes across the cluster. All these are well-known Cassandra anti-patterns, and administrators have to be vigilant about preventing users from incurring them. Even if one is perfectly aware of correct usage patterns, it’s easy to lose track of things like the size of non-frozen collections.

The new framework allows operators to restrict how Cassandra is used by:

* Disabling certain features.
* Disallowing some specific values.
* Defining soft and hard limits to certain database magnitudes.

=== Configuring Guardrails

> Guardrails are defined as regular properties in the Cassandra configuration file, https://cassandra.apache.org/doc/latest/cassandra/configuration/cass_yaml_file.html[`cassandra.yaml`]. They look like:

Reviewer: This link goes to the project's documentation, so I wonder if it should be ``link:/doc/latest/assandra/configuration/cass_yaml_file.html[`cassandra.yaml`]``, without the `https://cassandra.apache.org` prefix:
```suggestion
Guardrails are defined as regular properties in the Cassandra configuration file, /doc/latest/cassandra/configuration/cass_yaml_file.html[`cassandra.yaml`]. They look like:
```

Reviewer 2: Fixed in a separate commit. 🍻

```
tables_warn_threshold: -1
tables_fail_threshold: -1
secondary_indexes_per_table_warn_threshold: -1
secondary_indexes_per_table_fail_threshold: -1
allow_filtering_enabled: true
partition_keys_in_select_warn_threshold: -1
partition_keys_in_select_fail_threshold: -1
collection_size_warn_threshold:
collection_size_fail_threshold:
```

> Note that this is not an exhaustive list of all the available guardrails. There are many more, and new ones are under development, but this does give you an idea of the potential options. Note also that all guardrails are disabled by default. When enabled, a guardrail configuration might resemble the following: 

Reviewer 3: @dtopdontstop  it would be nice to inform a reader what he should do if he has an idea about a guardrail to be implemented.

Reviewer: That's an interesting idea, I guess we could add this phrase at the very end of the text:
```
If you have an idea for a new guardrail, please be free to open a https://issues.apache.org/jira/projects/CASSANDRA[JIRA] ticket or contact either on the https://lists.apache.org/list.html?dev@cassandra.apache.org[mail list] or https://s.apache.org/slack-invite[Slack channel].
```
However, if I'm understanding [the blog content process](https://cwiki.apache.org/confluence/display/CASSANDRA/Project+Blog+Content+Process) we are past the content review and this PR is only for formatting [the original doc](https://docs.google.com/document/d/1jV2diYUCaR1OED1zEnXUdLwFQpChuH_chz33IAfMO9I/edit).

Reviewer 2: Great idea @smiklosovic . I'm going to add a paragraph at the end of the blog encouraging users to log a ticket if they have ideas/proposals. /cc @adelapena 👍

```
tables_warn_threshold: 5
tables_fail_threshold: 10
secondary_indexes_per_table_warn_threshold: 5
secondary_indexes_per_table_fail_threshold: 10
allow_filtering_enabled: false
partition_keys_in_select_warn_threshold: 10
partition_keys_in_select_fail_threshold: 20
collection_size_warn_threshold: 10MiB
collection_size_fail_threshold: 20MiB
```

The guardrails defined in `cassandra.yaml` are applied as the node starts. It’s also possible to dynamically update the guardrails configuration through JMX at runtime. All guardrails are grouped under the MBean named `org.apache.cassandra.db.Guardrails`. There are plans to also support dynamically updating guardrails through virtual tables, although this option is not yet available.

=== Guardrails in Action

> Most guardrails are checked at the https://cassandra.apache.org/doc/latest/cassandra/cql/index.html[CQL layer], without involving the https://cassandra.apache.org/doc/latest/cassandra/architecture/storage_engine.html[storage engine] nor https://cassandra.apache.org/doc/latest/cassandra/architecture/dynamo.html[additional replicas]. Boolean guardrails for disabling features, such as `allow_filtering_enabled`, abort the operations attempting to use the disabled feature. For example, if the boolean guardrail for queries using filtering is disabled (`allow_filtering_enabled: false`) we will see a failure every time we try to run one of those queries, and the query won’t run:

Reviewer: I think we should use relative paths for the links to our own doc:
```suggestion
Most guardrails are checked at the /doc/latest/cassandra/cql/index.html[CQL layer], without involving the doc/latest/cassandra/architecture/storage_engine.html[storage engine] nor /doc/latest/cassandra/architecture/dynamo.html[additional replicas]. Boolean guardrails for disabling features, such as `allow_filtering_enabled`, abort the operations attempting to use the disabled feature. For example, if the boolean guardrail for queries using filtering is disabled (`allow_filtering_enabled: false`) we will see a failure every time we try to run one of those queries, and the query won’t run:
```

Reviewer 2: @adelapena you're correct. I've corrected them with `link:`s to the corresponding `*.adoc` pages. 🍻

=== Users and Extensibility

Guardrails are only applied to the operations of regular users, so they will neither be checked for superuser queries nor internal queries. By default, all regular users are subjected to the same guardrails configuration values defined in `cassandra.yaml`. However, the configuration for guardrails is an extensible API defined by the interfaces `GuardrailsConfig` and `GuardrailsConfigProvider`. These interfaces provide the configurations of every guardrail as a function of the user running the guarded operation. Third-party alternative implementations could provide different guardrail configurations depending on the user, or on some other factors. Custom implementations can be provided through the JVM system property `cassandra.custom_guardrails_config_provider_class`. It is important to know that this API isn’t officially supported yet, and there can be changes breaking backward compatibility in any minor release.

=== Background Guardrails

> In general, guardrails are associated with a specific CQL query. However, due to technical limitations, some guardrails are checked in the background, without being associated with any specific query. That’s the case, for example, with the guardrails for the size or number of items of a https://cassandra.apache.org/doc/trunk/cassandra/cql/types.html#collections[non-frozen collection]. Although we can do some checks when a query writes a new collection fragment, we cannot know if there are other fragments of the collection previously stored on the https://cassandra.apache.org/doc/trunk/cassandra/architecture/storage_engine.html#sstables[SSTables]. We could, of course, check into the SSTables, but that would involve a costly read-before-write operation. Instead, the guardrail checks the size of all collections every time an SSTable is written to disk, such as on https://cassandra.apache.org/doc/trunk/cassandra/architecture/storage_engine.html#memtables[memtable] flush or during https://cassandra.apache.org/doc/trunk/cassandra/operating/compaction/index.html[compaction]. If a large collection is detected the guardrail is triggered and emits the proper log messages and diagnostic events. However, in contrast to what happens with regular guardrails, we won’t abort any operation. The reason for this is that SSTable writes aren't associated with any specific CQL query. Rather, they are the result of multiple past CQL writes that have already happened long before. Future guardrails for similar things, like partition size, are likely to work in the same way.

Reviewer 2: Replace hardcoded `/doc` URLs with `link:`:
```suggestion
In general, guardrails are associated with a specific CQL query. However, due to technical limitations, some guardrails are checked in the background, without being associated with any specific query. That’s the case, for example, with the guardrails for the size or number of items of a link:/doc/trunk/cassandra/cql/types.html#collections[non-frozen collection^]. Although we can do some checks when a query writes a new collection fragment, we cannot know if there are other fragments of the collection previously stored on the link:/doc/trunk/cassandra/architecture/storage_engine.adoc#sstables[SSTables^]. We could, of course, check into the SSTables, but that would involve a costly read-before-write operation. Instead, the guardrail checks the size of all collections every time an SSTable is written to disk, such as on link:/doc/trunk/cassandra/architecture/storage_engine.html#memtables[memtable^] flush or during link:/doc/trunk/cassandra/operating/compaction/index.adoc[compaction^]. If a large collection is detected the guardrail is triggered and emits the proper log messages and diagnostic events. However, in contrast to what happens with regular guardrails, we won’t abort any operation. The reason for this is that SSTable writes aren't associated with any specific CQL query. Rather, they are the result of multiple past CQL writes that have already happened long before. Future guardrails for similar things, like partition size, are likely to work in the same way.
```

Reviewer: Oh, I missed the `link:` prefix on my suggestions about relative paths above.

At this time, there are guardrails for:

* Number of user keyspaces.
* Number of user tables.
* Number of columns per table.
* Number of secondary indexes per table.
* Number of materialized tables per table.
* Number of fields per user-defined type.
* Number of items in a collection.
* Number of partition keys selected by an IN restriction.
* Number of partition keys selected by the cartesian product of multiple IN restrictions.
* Allowed table properties.
* Allowed read consistency levels.
* Allowed write consistency levels.
* Collections size.
* Query page size.
* Minimum replication factor.
* Data disk usage, defined either as a percentage or as an absolute size.
* Whether user-defined timestamps are allowed.
* Whether GROUP BY queries are allowed.
* Whether the creation of secondary indexes is allowed.
* Whether the creation of uncompressed tables is allowed.
* Whether querying with ALLOW FILTERING is allowed.
* Whether dropping or truncating a table is allowed.

> It is worth mentioning that many of these guardrails were added in the last few months, some of them by newcomers to the project. That, in my opinion, indicates how easy it is to add new guardrails, and we expect to see many more guardrails in the future.

Reviewer 2: @smiklosovic adding your idea at the end here (/cc @adelapena):

```suggestion
It is worth mentioning that many of these guardrails were added in the last few months, some of them by newcomers to the project. That, in my opinion, indicates how easy it is to add new guardrails, and we expect to see many more guardrails in the future.

If you have ideas for new guardrails, please feel free https://issues.apache.org/jira/secure/CreateIssue.jspa?pid=12310865&issuetype=4[to log a ticket^] with details of your proposal and we would be happy to look into it.
```
